### PR TITLE
Add `addDefaultsToDescriptions`

### PR DESCRIPTION
**Pull request by herbertgithubaccount**
## Summary

- This method will recursively add a string to the description of each `Setting` variable that displays the default value of that setting.
- The default value is formatted using the `Setting:convertToLabelValue` method, so it is immediately compatible with every `Setting`, and it respects custom user-formatting.
- The default value is retrieved by first checking for `variable.defaultValue`. If `variable.defaultValue` is nil, the method will scan through a provided `defaultConfig`.
- This method can also be called on a `(SideBar)Page`, `Category`, or `Template`. Doing so will result in the method being propagated to all subcomponents.
  - This means it's possible to call the method once, on a `Template`, and have the default values displayed for the whole config menu.

This is a draft PR because I have not yet created the relevant documentation. And some minor tweaks may be needed. But I have tested the current implementation on my own MCM pages and things seem to work as intended.

## Detailed Description

This method works differently on different types of components. Here's what happens for each type of component:
- `Template`: propagate method call to everything in `self.pages`.
- `Category`: propagate method call to everything in `self.components`.
  - **Note:** This implementation gets inherited by `Page`, since it is a subclass of `Category`.
- `Setting`: 
  - Checks if it has a `self.variable` that is a `TableVariable`. Exits early if not.
  - Retrieves a `defaultValue`. 
    - This is done by first checking `self.variable.defaultSetting`, and then using a recursive call that fetches it from a `defaultConfig` parameter, if `self.variable.defaultSetting == nil`.
    - The recursive function will scan subtables of `defaultConfig` until it finds a `defaultTable` that is a subtable of `defaultConfig` (possibly with `defaultTable == defaultConfig`), with a `key` that matches `self.variable.id`. It will then do a heuristic check on the `defaultTable` to make sure this is the correct subtable. (This is to handle the case where two subtables have a key in common.)
    -  If the heuristic check fails, this recursive function will keep scanning subtables until a better match is found.
  - After `defaultValue` has been retrieved, it gets appended to the setting's `description` field. This works as follows:
    - A default description is generated that looks like `"Default: <defaultValue>"`.
      - This description makes use of `mwse.mcm.i18n` so that it's compatible with future translations. i.e., the description is really given by `string.format("%s: %s", mwse.mcm.i18n("Default"), defaultValue`.
    - If `self.description == nil`, then `self.description` will be set to `"Default: <defaultValue>"`.
    - If `self.description ~= nil`, then `self.description` will be set to `self.description .. "\n\n" .. "Default: <defaultValue>"`.

## Notes
- Documentation still needs to be generated.
- The current heuristic is not perfect. 
  - It will fail if `defaultConfig` has two identical subtables. But I do not know of a good way to prevent this. 
  - This issue can be bypassed in at least two ways:
    - The first method is to directly `Category:addDefaultsToDescriptions(defaultSubTable)` on each `Category` (or `Page`) that stores the relevant settings. (Assuming the relevant `Setting` components are stored in different `Categories` or `Pages`.)
    - The second approach would be to just use the `defaultValue` setting on the `TableVariable`'s in question.
  - But in any case, the  
- The current name for this method is very verbose, and I am open to suggestions on a better name.
- This method currently does not play nicely with `mwseMCMDropDown`, but that's because `mwseMCMDropdown` does not have a custom implementation of `convertToLabelValue`. I'm planning on addressing that in a separate PR.

### CODE REVIEW

File: misc/package/Data Files/MWSE/core/mcm/components/Component.lua
Comment (C3pa) [inline misc/package/Data Files/MWSE/core/mcm/components/Component.lua:246]: Is it needed to declare the method on the parent `Component` class? To me, it looks like this can be declared the Category component.

File: misc/package/Data Files/MWSE/core/mcm/components/settings/Setting.lua
Comment (C3pa) [inline misc/package/Data Files/MWSE/core/mcm/components/settings/Setting.lua:122]: After some discussion with Merlord (which you can see on our Discord channel), I think it would be best not to do the heuristic finding of the default values from the `defaultConfig` table. Only use the `defaultSetting` field provided by the variable.

File: misc/package/Data Files/MWSE/core/mcm/components/templates/Template.lua
Comment (C3pa) [inline misc/package/Data Files/MWSE/core/mcm/components/templates/Template.lua:365]: Can you instead of making this a method in `Template`, be behind a flag on `Template:new` method, and if it's set to `true` call the `addDefaultsToDescriptions` method on all of its pages and also call that method in this snippet:
```lua
	Template[key] = function(self, data)
		if not data then
			data = {}
		elseif type(data) == "string" then
			data = { label = data }
		end
		data.parentComponent = self
		local component = mwse.mcm[key](data)
		table.insert(self.pages, component)
		return component
	end
```

**Review by C3pa — CHANGES_REQUESTED** [review]
After a discussion with Merlord, I think it would be best not to do the heuristic lookup of default values. Only use the default values given explicitly by the `defaultSetting` field of the variable.

File: misc/package/Data Files/MWSE/core/mcm/components/Component.lua
Comment (herbertgithubaccount) [reply misc/package/Data Files/MWSE/core/mcm/components/Component.lua:246]: I'm not really sure that this is the best solution. I just found it made the implementation a bit easier since it means I didn't have to add a check to make sure `component.componetType` is a category or setting or page before calling the method. But I think that the ability to call it on a component probably should not be included in the documentation, and just kept as a secret implementation detail.

But maybe it's better to just explicitly check before calling it each time?

**Comment by herbertgithubaccount** [comment]
I removed the `defaultConfig` parameter so that this PR works better with the changes proposed in #546 . I haven't tested this new version yet, but I wanted to wait until #546 gets merged before I start testing. The documentation needs to be built before this is merged anyways.

**Comment by C3pa** [comment]
Can you update this branch from master?

**Comment by C3pa** [comment]
Here is a conceptual downside I see if `addDefaultsToDescriptions` is a method on Template. Let's consider:
```lua
local defaultConfig = {
	a = 1, b = 2, c = 3, d = 4,
	subConfig = { e = 5, f = 6},
}

local config = mwse.loadConfig("A test mod", defaultConfig)

event.register("modConfigReady", function()	
	local template = mwse.mcm.createTemplate{
		label = "Experimental MCM",
	}
	template:saveOnClose("A test mod", config)
	template:register()

	-- Doesn't add anything to the description. No pages/settings have been created yet, so there are no descriptions yet.
	-- Still, looks like it would be a common pitfall since it sounds logical you'd want to do all the work you have
	-- with your template then move on to add the pages and settings.
	-- template:addDefaultsToDescriptions()

	local page = template:createSideBarPage{label = "Page"}
	page:createSlider{
		label = "a", 
		variable = mwse.mcm.createTableVariable({ id = "a", table = config, defaultSetting = defaultConfig.a }),
	}
	page:createSlider{
		label = "b", 
		variable = mwse.mcm.createTableVariable({ id = "b", table = config, defaultSetting = defaultConfig.b }),
	}
	page:createSlider{
		label = "c", 
		variable = mwse.mcm.createTableVariable({ id = "c", table = config, defaultSetting = defaultConfig.c }),
	}
	page:createSlider{
		label = "d", 
		variable = mwse.mcm.createTableVariable({ id = "d", table = config, defaultSetting = defaultConfig.d }),
	}

	local category = page:createCategory{label = "subConfig", configKey = "subConfig" }
	category:createSlider{
		label = "e", 
		variable = mwse.mcm.createTableVariable({ id = "e", table = config.subConfig, defaultSetting = defaultConfig.subConfig.e }),
	}
	category:createSlider{
		label = "f",
		variable = mwse.mcm.createTableVariable({ id = "f", table = config.subConfig, defaultSetting = defaultConfig.subConfig.f }),
	}

        -- Finally, this is the place we have to call `addDefaultsToDescriptions`.
	template:addDefaultsToDescriptions()
end)
```

What I wanted to see in [this comment](https://github.com/MWSE/MWSE/pull/542#discussion_r1650028285) was mostly what I did in #547. By making this an argument in Template's constructor, we just pass a flag there and we get default text added to all the descriptions:
```lua
local template = mwse.mcm.createTemplate{
		label = "Experimental MCM", 
		addDefaultsToDescriptions = true,
}
```

**Comment by herbertgithubaccount** [comment]
Hmmm that is a very good point. I can see how it would be counterintuitive to have to put it at the bottom. Making it a parameter to the constructor does like the better option.

Here are a couple potential points of tension that might come from that, and then later on I'll present a potential solution to these:
- It may also be a bit confusing that it's a parameter to the `Template` constructor, but not to the `Page`/`Category`/`Setting` constructors.
- If it's a parameter to the constructor, it may not play nicely with the description being updated after the setting is created. Whereas, if it were a method, then you could update the description of a `setting` and then call `setting:addDefaultsToDescription()`. (But this approach is also a clunky and counterintuitive.)

How would you feel about this solution:
- Implement `addDefaultsToDescriptions` as a parameter to the constructor of `Template`, `Category`, and `Setting`, and make it "propagate downwards", like the `config` parameter in #546. (i.e., you can specify it for `Template`, and it will be inherited by all `Categories`, or you can specify it for a `Category`, and that will be inherited by all `Setting`s in that category.)
- Make this implementation alter the description that's shown in the sidebar page, but _not_ the description that's stored in the MCM component. This second idea would have two other advantages
  - It wouldn't mess up the generated descriptions if someone were to do things like `description = string.format("%s %s", description, somethingElse)`, later on.
  - It would handle any changes to the description in a seamless way.

I think this solution could be implemented by changing `MouseOverInfo:updateInfo`, maybe to be something like:
```lua
--- @param component mwseMCMComponent|nil
function MouseOverInfo:updateInfo(component)
    -- If component has a description, update mouseOver
    -- Or return to original text on mouseLeave
    local newText
    if component then
        -- Try to add default description
        local var = component.variable
        if component.addDefaultsToDescriptions and var and var.defaultSetting then
            local defaultStr = component:convertToLabelValue(var.defaultSetting)
            if component.description then    
                newText = string.format(
                    "%s\n\n%s: %s", 
                    self.description, mwse.mcm.i18n("Default"), defaultStr
                )
            else
                newText = string.format("%s: %s", mwse.mcm.i18n("Default"), defaultStr)
            end
        end
        -- Default to component description
        newText = newText or component.description
    end
    -- Default to `MouseOverInfo` description.
    newText = newText or self.text or ""
    self.elements.info.text = newText
    self:update()
end
```

(I should mention that I'm not happy with the nesting level or general niceness of the above implementation, it's more-so just a rough sketch of what the functionality could look like, if we decide to go this route.)

What do you think of this approach? If you think this could be a good way forward, I can start working on an implementation that's more tested and nicer to look at.

**Comment by C3pa** [comment]
> How would you feel about this solution

I really like the idea. Actually, I think it's the best of already proposed things since the description in MCM only makes sense for SideBarPage's sidebar (MouseOverInfo). Moreover, not modifying the original `description` fields of settings is a bonus. Go for it.

Just a logistical note, I guess it makes sense to start working on the new approach on a different branch, based on master, since it doesn't have much in common with the approach in this PR.

**Pull request closed without merging**
